secrets/provider/kubernetes/provider.go

```diff
@@ -707,32 +741,34 @@ func rulesForSecretAccess(
 func (k *kubernetesClient) ensureBindingForSecretAccessToken(
 	ctx context.Context, sa *core.ServiceAccount, owned, read, removed []string,
 ) (cleanups []func(), _ error) {
-	role, err := k.getRole(ctx, sa.Name)
-	if err != nil && !errors.Is(err, errors.NotFound) {
-		return cleanups, errors.Trace(err)
-	}
-	if errors.Is(err, errors.NotFound) {
-		_, err = k.createRole(ctx,
-			&rbacv1.Role{
-				ObjectMeta: v1.ObjectMeta{
-					Name:        sa.Name,
-					Namespace:   k.namespace,
-					Labels:      sa.Labels,
-					Annotations: sa.Annotations,
-				},
-				Rules: rulesForSecretAccess(k.namespace, false, nil, owned, read, removed),
+	role, err := k.createRole(ctx,
+		&rbacv1.Role{
+			ObjectMeta: v1.ObjectMeta{
+				Name:        sa.Name,
+				Namespace:   k.namespace,
+				Labels:      sa.Labels,
+				Annotations: sa.Annotations,
 			},
-		)
-		if err == nil {
-			cleanups = append(cleanups, func() { _ = k.deleteRole(ctx, role.GetName(), role.GetUID()) })
+			Rules: rulesForSecretAccess(k.namespace, false, nil, owned, read, removed),
+		},
+	)
+	if errors.Is(err, errors.AlreadyExists) {
+		role, err = k.getRole(ctx, sa.Name)
+		if err != nil {
+			return cleanups, errors.Annotatef(err, "getting role %q", sa.Name)
 		}
-	} else {
 		role.Rules = rulesForSecretAccess(k.namespace, false, role.Rules, owned, read, removed)
 		_, err = k.updateRole(ctx, role)
+		if err != nil {
+			return cleanups, errors.Annotatef(err, "updating role %q", sa.Name)
+		}
 	}
 	if err != nil {
-		return cleanups, errors.Trace(err)
+		return cleanups, errors.Annotatef(err, "creating role %q", sa.Name)
 	}
+
+	cleanups = append(cleanups, func() { _ = k.deleteRole(ctx, role.GetName(), role.GetUID()) })
```

Comment:
This needs better structure, use else blocks to clearly define the logic. The current logic will add a clean up step for a role it updated (not created).

```go
if errors.Is(err, errors.AlreadyExists) {
	role, err = k.getRole(ctx, sa.Name)
	if err != nil {
		return cleanups, errors.Annotatef(err, "getting role %q", sa.Name)
	}
	role.Rules = rulesForSecretAccess(k.namespace, false, role.Rules, owned, read, removed)
	_, err = k.updateRole(ctx, role)
	if err != nil {
		return cleanups, errors.Annotatef(err, "updating role %q", sa.Name)
	}
} else if err != nil {
	return cleanups, errors.Annotatef(err, "creating role %q", sa.Name)
} else {
	cleanups = append(cleanups, func() { _ = k.deleteRole(ctx, role.GetName(), role.GetUID()) })
}
```

Reply:
refactored w thanks